Sources/SourceKitLSP/Swift/ReferenceDocumentURL.swift

```diff
@@ -0,0 +1,75 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LanguageServerProtocol
+
+/// A Reference Document is a document whose url scheme is `sourcekit-lsp:` and whose content can only be retrieved
+/// using `GetReferenceDocumentRequest`. The enum represents a specific type of reference document and its
+/// associated value represents the data necessary to generate the document's contents and its url
+///
+/// The `url` will be of the form: `sourcekit-lsp://<document-type>/<display-name>?<parameters>`
+/// Here,
+///  - The `<document-type>` denotes the kind of the content present in the reference document
+///  - The `<parameters>` denotes the parameter-value pairs such as "p1=v1&p2=v2&..." needed to generate
+/// the content of the reference document.
+///  - The `<display-name>` is the displayed file name of the reference document. It doesn't involve in generating
+/// the content of the reference document.
+package enum ReferenceDocumentURL {
+  case macroExpansion(MacroExpansionReferenceDocumentURLData)
+
+  var url: URL {
+    get throws {
+      switch self {
+      case let .macroExpansion(data):
+        return try data.url
+      }
+    }
+  }
+
+  init(from uri: DocumentURI) throws {
+    guard let url = URL(string: uri.stringValue) else {
```

Comment:
Instead of going through the `stringValue` and then parsing that as a `URL` again, we should expose `DocumentURI`. Since in almost all use cases users want `fileURL`, I would name it something that sounds a little scary, like

```swift
/// The URL representation of the URI. Note that this URL can have an arbitrary scheme and might not represent a file URL.
public var arbitrarySchemURL: URL { storage }
```